src/main/java/fi/metatavu/keycloak/twilio/sendgrid/otp/authenticator/EmailOtpAuthenticatorForm.java

```diff
@@ -0,0 +1,194 @@
+package fi.metatavu.keycloak.twilio.sendgrid.otp.authenticator;
+
+import com.sendgrid.Method;
+import com.sendgrid.Request;
+import com.sendgrid.SendGrid;
+import com.sendgrid.helpers.mail.Mail;
+import com.sendgrid.helpers.mail.objects.Content;
+import com.sendgrid.helpers.mail.objects.Email;
+import lombok.extern.jbosslog.JBossLog;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.keycloak.authentication.*;
+import org.keycloak.common.util.SecretGenerator;
+import org.keycloak.events.Errors;
+import org.keycloak.forms.login.LoginFormsProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.FormMessage;
+import org.keycloak.services.messages.Messages;
+
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.List;
+
+@JBossLog
+public class EmailOtpAuthenticatorForm implements Authenticator {
+
+    static final String ID = "sendgrid-email-otp-form";
```

Comment:
Could this also be generalized to email-otp-form? Code can still use sendgrid, but we can add later other providers for sending emails